File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one.java
class Solution {
    public int numSteps(String s) {
        int cnt = 0;
        while(!s.equals("1")){
            if(s.charAt(s.length()-1)=='1'){
                s = add(s);
                cnt++;
            }
            else{
                s = s.substring(0, s.length()-1);
                cnt++;
            }
        }

        return cnt;
    }

    public String add(String s){
        int carry = 1;
        String res = "";
        for(int i = s.length()-1; i>=0; i--){
            if(s.charAt(i)=='1' && carry==1){
                res = '0'+res;
            }
            else if((s.charAt(i)=='1' && carry == 0) || (s.charAt(i)=='0' && carry == 1)){
                res = '1'+res;
                carry = 0;
            }
            else{
                res = '0'+res;
            }
        }

        if(carry==1){
            res = '1'+res;
        }

        return res;
    }
}

File: 1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one/1404-number-of-steps-to-reduce-a-number-in-binary-representation-to-one.py
class Solution:
    def numSteps(self, s: str) -> int:
        cnt = 0

        while(s!="1"):
            if(s[len(s)-1]=="1"):
                s = self.add(s)
                cnt+=1
            else:
                s = s[:len(s)-1]
                cnt+=1
        
        return cnt

    def add(self, s):
        carry = 1
        res = ""
        for i in range(len(s)-1, -1, -1):
            if((s[i]=="1" and carry==1) or (s[i]=="0" and carry==0)):
                res = "0"+res
            else:
                res = "1"+res
                carry = 0
        
        if(carry==1):
            res = "1"+res

        return res        
        